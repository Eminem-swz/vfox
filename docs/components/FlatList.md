# FlatList 回收列表

注：

- 需要给 `flat-list` 一个固定高度，通过 CSS 设置 height。
- 如果列表内容存在不固定宽高的图片，由于图片加载的机制，每个 item 的 DOM 渲染了，图片可能还未加载，导致每个 item 的位置计算错误，所以作为调用方，需要将未加载的图片用样式固定住。在已知宽高比（比如常见的正方形商品图），可以使用 `Image` 组件内置 `aspect-ratio` 来固定图片宽高。

## Import

```
import { FlatList } from 'vfox'
```

具体的引入方式可以参考[引入组件](../guide/import.md)。

## Props

| 属性                  | 类型            | 默认值 | 必填 | 说明                                                                                                                                                                   |
| --------------------- | --------------- | ------ | ---- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------- |
| data                  | any[]           |        | 是   | 列表数组                                                                                                                                                               |
| data-key              | string          |        | 否   | 表明每个 item 项的身份（item[dataKey]），如果想指向自己，则设置为 `*this` 。建议填写，不填默认使用 index 作为身份标识                                                  |
| horizontal            | boolean         | false  | 否   | 设置为 true 则变为水平布局模式                                                                                                                                         |
| get-item-size         | Function        |        | 否   | 动态返回列表项尺寸（垂直布局下指高度，水平布局下指宽度），`getItemSize(item:any, index:number) => number`，item 是 data[index] 的副本                                  |
| item-size             | number          |        | 否   | 设置列表项尺寸，优先使用 `get-item-size`                                                                                                                               |
| end-reached-threshold | number          | 0.5    | 否   | 决定当距离内容最底部还有多远时触发 onEndReached 回调。注意此参数是一个比值而非像素单位。比如，0.5 表示距离内容最底部的距离为当前列表可见长度的一半时触发               |
| enable-pull-refresh   | boolean         | false  | 否   | 是否开启下拉刷，如果时水平列表则为左拉刷新，搭配 `refreshing` 事件使用新                                                                                               |
| lower-loading         | boolean         | false  | 否   | 是否展示底部加载更多状态，配合 `end-reached` 事件加载无无限列表                                                                                                        |
| initial-waterfall     | boolean         | false  | 否   | 是否开启瀑布流展示方，仅初始化时生效式                                                                                                                                 |
| waterfall-col-count   | number          | 2      | 否   | 瀑布流列数，`initial-waterfall=true` 时生效，支持 2 ~ 5 列                                                                                                             |
| item-gutter           | number/number[] | false  | 否   | 项间隔，通过数值设置水平和垂直间隔，也可以通过数组的形式分别设置水平和垂直间隔 [水平间隔, 垂直间隔]，`getItemSize` 或者 `itemSize` 设定值需要把 `item-gutter` 考虑进去 |

## Events

| 事件           | 描述                 | 回调函数参数                                                                                           |
| -------------- | -------------------- | ------------------------------------------------------------------------------------------------------ |
| recycle-change | 列表项回收变化时触发 | { item, index, recycled }                                                                              |
| end-reached    | 滚动到末尾时触发     | { distanceFromEnd } 其中 distanceFromEnd 为距离末尾的距离，单位 px                                     |
| scroll         | 滚动时触发           | { scrollLeft, scrollTop, scrollHeight, scrollWidth }                                                   |
| refreshing     | 下拉刷新时触发       | ({ pullDirection: string }, done: Function) 其中 pullDirection 指下拉的方向，done 指刷新完毕回调的函数 |

### recycle-change 的参数

| 值       | 类型    | 说明                                           |
| -------- | ------- | ---------------------------------------------- |
| recycled | boolean | 是否被回收                                     |
| index    | number  | 第 index 项                                    |
| item     | any     | 传入 data[index]的副本，修改不会影响 prop data |

## Slots

### 列表项（#default）

```
<fx-flat-list>
 <template #default="{ item, index }">
  {{ index }} : {{ item }}
 </template>
</fx-flat-list>
```

### 列表为空（#empty）

```
<fx-flat-list>
  <template #empty>暂无数据</template>
  ...
</fx-flat-list>
```

### 分割线（#separator）

```
<fx-flat-list>
  <template #separator>
    <div class="line"></div>
  </template>
</fx-flat-list>
```

注：`getItemSize` 或者 `itemSize` 设定值需要把分割线也考虑进去。

### 前置（#header）

```
<fx-flat-list>
  <template #header></template>
  ...
</fx-flat-list>
```

也可以传入 [Icon](./Icon.md)，比如常见的搜索。

### 后置（#footer）

```
<fx-flat-list>
  ...
  <template #footer>暂时没有更多了</template>
</fx-flat-list>
```

## Methods

| 方法名            | 说明                                                 | 参数                                                                       |
| ----------------- | ---------------------------------------------------- | -------------------------------------------------------------------------- |
| scrollToIndex     | 将位于指定位置的元素滚动到可视区的指定位置           | ({ index: number, animated: boolean, viewPosition: ViewPosition }) => void |
| scrollTo          | 滚动列表到指定的偏移，单位 px                        | ({ offset: number, animated: boolean }) => void                            |
| scrollToEnd       | 滚动到底部                                           | ( animated: boolean ) => void                                              |
| recordInteraction | 主动通知列表发生了一个事件，以使列表重新计算可视区域 | () => void                                                                 |

### scrollToIndex 的参数

| 属性         | 类型    | 默认值  | 必填 | 说明                                                          |
| ------------ | ------- | ------- | ---- | ------------------------------------------------------------- |
| index        | number  |         | 是   | 列表第 `index` 项滚动到可视区的指定位置                       |
| animated     | boolean | true    | 否   | 滚动过程中是否使用过度动画                                    |
| viewPosition | string  | 'start' | 否   | 'start'/'center'/'end'/0/0.5/1 显示在屏幕的头部/中间/末尾位置 |
