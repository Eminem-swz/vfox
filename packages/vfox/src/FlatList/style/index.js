import '../../style/common/style/index'
import '../../ScrollView/style/index'
import '../index.scss'
