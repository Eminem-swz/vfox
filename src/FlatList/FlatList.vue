<template>
  <ScrollView
    class="fx-flat-list"
    :class="{ horizontal }"
    :scrollX="scrollX"
    :scrollY="scrollY"
    :lowerThreshold="lowerThreshold"
    :enablePullDirections="enablePullDirections"
    :lowerLoading="lowerLoading"
    @scroll="onScroll"
    @scrollToLower="onScrollToLower"
    @refreshing="onRefreshing"
    ref="scrollView"
  >
    <div class="fx-flat-list_header" v-if="$slots.header">
      <slot name="header"></slot>
    </div>
    <ul class="fx-flat-list_list" ref="listEl">
      <li
        class="fx-flat-list_item"
        v-for="(item, index) in list"
        :key="
          dataKey
            ? dataKey === '*this'
              ? item.data
              : item.data[dataKey]
            : index
        "
      >
        <div
          class="fx-flat-list_item-inner"
          v-show="!item.recycled"
          :style="itemStyles"
        >
          <slot :item="item.data" :index="index"> </slot>
        </div>
        <div
          class="fx-flat-list_separator"
          v-if="cols.length <= 1 && $slots.separator && index < list.length - 1"
        >
          <slot name="separator"></slot>
        </div>
      </li>
    </ul>
    <LoadMore class="fx-flat-list_indicator" v-show="lowerLoading" loading>
      {{ locale.flatListLoadingText }}
    </LoadMore>
    <div class="fx-flat-list_empty" v-show="list.length === 0">
      <slot name="empty"></slot>
    </div>
    <div class="fx-flat-list_footer" v-if="$slots.footer">
      <slot name="footer"></slot>
    </div>
  </ScrollView>
</template>

<script lang="ts">
import { computed, defineComponent, onMounted, ref, nextTick, watch } from 'vue'
import type { PropType } from 'vue'
import { LoadMore } from '../LoadMore'
import { ScrollView } from '../ScrollView'
import {
  cloneData,
  isInNumberRange,
  isInteger,
  isNumber,
  rangeInteger
} from '../helpers/util'
import Exception from '../helpers/exception'
import { getRelativeOffset } from '../helpers/dom'
import { useResizeDetector } from '../hooks/use-resize-detector'
import type { StyleObject, ViewPosition } from '../helpers/types'
import type { ScrollToIndexOptions } from './types'
import { locale } from '../Locale'
import type {
  OnRefreshing,
  OnScroll,
  PullDirection,
  ScrollToOptions
} from '../ScrollView/types'
import {
  emitRefreshingValidator,
  emitScrollValidator
} from '../ScrollView/scrollView'
import { parseGutter, propGutter } from '../Row/row'

interface FlatItemElement extends HTMLElement {
  _recycled?: boolean
  _translateOffset?: number
}

type Item = any

export default defineComponent({
  name: 'fx-flat-list',
  components: { ScrollView, LoadMore },
  props: {
    dataKey: {
      type: String,
      default: null
    },
    data: {
      type: Array as PropType<Item[]>,
      required: true,
      default: () => [] as Item[]
    },
    getItemSize: {
      type: Function,
      default: null
    },
    itemSize: {
      type: Number,
      default: null
    },
    // 横向
    initialHorizontal: {
      type: Boolean,
      default: false
    },
    // 预加载多少屏
    preLoad: {
      type: Number,
      default: 2
    },
    // 决定当距离内容最底部还有多远时触发endReached回调
    endReachedThreshold: {
      type: Number,
      default: 0.5
    },
    // 是否开启下拉刷新
    enablePullRefresh: {
      type: Boolean,
      default: false
    },
    lowerLoading: {
      type: Boolean,
      default: false
    },
    initialWaterfall: {
      type: Boolean,
      default: false
    },
    waterfallColCount: {
      type: Number,
      validator: (val: number) => {
        return isInteger(val) && isInNumberRange(val, 2, 5)
      },
      default: 2
    },
    itemGutter: propGutter
  },
  emits: {
    'recycle-change': (payload: {
      recycled: boolean
      index: number
      item: Item
    }) =>
      payload &&
      typeof payload.recycled === 'boolean' &&
      typeof payload.index === 'number',
    'end-reached': (payload: { distanceFromEnd: number }) =>
      payload && typeof payload.distanceFromEnd === 'number',
    scroll: emitScrollValidator,
    refreshing: emitRefreshingValidator
  },
  setup(props, { emit }) {
    const cols = ref<number[]>([])
    const list = ref<
      {
        data: Item
        recycled: boolean
      }[]
    >([])
    const scrollView = ref<InstanceType<typeof ScrollView>>()
    const listEl = ref<HTMLElement>()

    let wrapperSize = 0
    let scrollX = false
    let scrollY = true
    let scrollCount = 0
    let horizontal = false
    let colSize = 0

    if (props.initialWaterfall) {
      for (
        let i = 0, len = rangeInteger(props.waterfallColCount, 2, 5);
        i < len;
        i++
      ) {
        cols.value.push(0)
      }
    } else if (props.initialHorizontal) {
      horizontal = true
    }

    if (horizontal) {
      scrollX = true
      scrollY = false
    }

    function isSizeFixed() {
      return props.getItemSize != null || props.itemSize != null
    }

    function dataToList(data: Item[]) {
      const newList: Item[] = []
      const dataKey = props.dataKey

      data.forEach((v, k) => {
        const oldItem = list.value[k]
        // 如果没有固定高度（或宽度），默认不回收，因为需要展示以获取当前高度
        const newItem = { data: v, recycled: isSizeFixed() }

        if (oldItem) {
          if (dataKey === '*this' && v === oldItem.data) {
            newItem.recycled = oldItem.recycled
          } else if (dataKey && v[dataKey] === oldItem.data[dataKey]) {
            newItem.recycled = oldItem.recycled
          } else if (!dataKey) {
            newItem.recycled = oldItem.recycled
          }
        }

        newList.push(newItem)
      })

      list.value = newList
    }

    function getRootEl() {
      return scrollView.value?.$el as HTMLElement
    }

    function updateSize() {
      const $root = getRootEl()

      wrapperSize = getElSize($root)
      colSize = $root.offsetWidth / cols.value.length
    }

    function getElSize($el: HTMLElement) {
      return $el[scrollX ? 'offsetWidth' : 'offsetHeight']
    }

    function getScrollSize() {
      return getRootEl()[scrollX ? 'scrollLeft' : 'scrollTop']
    }

    const onRefreshing: OnRefreshing = (res, loadComplete) => {
      emit('refreshing', res, loadComplete)
    }

    /**
     * 重新计算列表元素
     * @param scrollSize 滚动值
     * @param source 调用来源
     */
    function updateItems(scrollSize: number | null, source?: string) {
      // console.log('updateItems', source)

      const $list = listEl.value as HTMLElement
      const newCols = cols.value.map(() => {
        return 0
      })

      const _scrollSize =
        scrollSize == null ? getScrollSize() : (scrollSize as number)

      getItemEls().forEach(($item, index) => {
        let offset = $list.offsetTop
        const sizeKey = scrollX ? 'width' : 'height'
        const itemLayout = getItemLayout($item, index)

        if (itemLayout.fixed) {
          $item.style[sizeKey] = itemLayout.size + 'px'
        } else {
          $item.style[sizeKey] =
            $item[scrollX ? 'offsetWidth' : 'offsetHeight'] + 'px'
        }

        if (newCols.length > 1) {
          const colMin = Math.min.apply(null, newCols)
          const colMinIndex = newCols.indexOf(colMin)

          $item.style.width = colSize + 'px'
          $item.style.position = 'absolute'
          $item.style.transform = `translate3d(${
            colMinIndex * colSize
          }px, ${colMin}px, 0)`

          newCols[colMinIndex] = colMin + $item.offsetHeight
          $item._translateOffset = offset = colMin
        } else {
          offset = $item[scrollX ? 'offsetLeft' : 'offsetTop']
        }

        const item = list.value[index]

        const change = (recycled: boolean) => {
          $item._recycled = recycled

          emit('recycle-change', {
            recycled,
            index,
            item: cloneData(item.data)
          })
        }

        if (
          offset >= _scrollSize - wrapperSize * props.preLoad &&
          offset <= _scrollSize + wrapperSize * (props.preLoad + 1)
        ) {
          // 展示
          // if (!itemLayout.fixed) {
          //   $item.style[sizeKey] = ''
          // }
          item.recycled = false

          if ($item._recycled == null || $item._recycled === true) {
            // 新的元素或者之前被回收的
            change(item.recycled)
          }
        } else {
          // 被回收
          // if (!itemLayout.fixed) {
          //   $item.style[sizeKey] = itemLayout.size + 'px'
          // }
          item.recycled = true

          if ($item._recycled == null || $item._recycled === false) {
            // 新的元素或者之前被回收的
            change(item.recycled)
          }
        }
      })

      if (newCols.length > 1) {
        $list.style.height = Math.max.apply(null, newCols) + 'px'
        cols.value = newCols
      }
    }

    /**
     * 获取列表元素
     */
    function getItemEls() {
      const $list = listEl.value as HTMLElement
      const $els: FlatItemElement[] = []

      if (!$list) {
        return $els
      }

      const $children = $list.children

      for (let i = 0, len = $children.length; i < len; i++) {
        const $el = $children[i]
        if ($el.className.indexOf('item') > -1) {
          $els.push($el as FlatItemElement)
        }
      }

      return $els
    }

    function getItemLayout($el: HTMLElement, index: number) {
      if (typeof props.getItemSize === 'function') {
        try {
          const size = props.getItemSize(cloneData(props.data[index]), index)

          if (isNumber(size)) {
            return {
              fixed: true,
              size
            }
          }
        } catch (error) {
          console.error(
            new Exception(
              'The object.size value returned by getItemSize should be a Number type.',
              Exception.TYPE.PROP_ERROR,
              'FlatList'
            )
          )
        }
      } else if (isNumber(props.itemSize)) {
        return {
          fixed: true,
          size: props.itemSize
        }
      }

      return {
        size: getElSize($el),
        fixed: false
      }
    }

    // /**
    //  * 设置列表项的尺寸，只在固定尺寸下生效
    //  */
    // function setItemsSize() {
    //   getItemEls().forEach(($item, index) => {
    //     const itemLayout = getItemLayout($item, index)

    //     if (itemLayout.fixed) {
    //       $item.style[scrollX ? 'width' : 'height'] = itemLayout.size + 'px'
    //     }
    //   })
    // }

    /**
     * 主动通知列表发生了一个事件，以使列表重新计算可视区域
     */
    function recordInteraction() {
      updateItems(null, 'recordInteraction')
    }

    function onResize() {
      updateSize()
      updateItems(null, 'resize')
    }

    let scrollTimer: number

    const onScroll: OnScroll = res => {
      const scrollSize = res[scrollX ? 'scrollLeft' : 'scrollTop']

      if (scrollCount > 10) {
        // 每轮询10次更新一次
        scrollCount = 0
        clearTimeout(scrollTimer)
        updateItems(scrollSize, 'scroll')
      } else {
        clearTimeout(scrollTimer)
        scrollTimer = window.setTimeout(() => {
          scrollCount = 0
          updateItems(scrollSize, 'scroll')
        }, 17)
        scrollCount++
      }

      // scroll 事件回调
      emit('scroll', res)
    }

    function onScrollToLower() {
      const $el = getRootEl()
      const distanceFromEnd = scrollX
        ? $el.scrollWidth - $el.scrollLeft - $el.offsetWidth
        : $el.scrollHeight - $el.scrollTop - $el.offsetHeight

      if (!props.lowerLoading) {
        emit('end-reached', {
          distanceFromEnd
        })
      }
    }

    /**
     * 滚动列表到指定的偏移（以像素为单位）
     */
    function scrollToOffset(options: number | ScrollToOptions) {
      scrollView.value?.scrollTo(options)
    }

    /**
     * 滑动到第index个元素
     */
    function scrollToIndex(options: number | ScrollToIndexOptions) {
      let index: number
      let animated = true
      let viewPosition: ViewPosition | undefined

      if (typeof options === 'number') {
        index = options
        options = {
          index
        }
      } else {
        index = options.index
        animated = !(options.animated === false)
        viewPosition = options.viewPosition
      }

      const $view = getItemEls()[index]

      if ($view) {
        const parentOffset = getRelativeOffset($view, getRootEl(), viewPosition)

        const offset =
          parentOffset[scrollX ? 'offsetLeft' : 'offsetTop'] +
          ($view._translateOffset || 0)

        scrollToOffset({
          offset,
          animated
        })
      }
    }

    /**
     * 滚动到底部
     */
    function scrollToEnd(animated = false) {
      const $root = getRootEl()

      if (!$root) {
        return
      }

      scrollToOffset({
        offset: scrollX ? $root.scrollWidth : $root.scrollHeight,
        animated
      })
    }

    const lowerThreshold = computed(() => {
      return wrapperSize * props.endReachedThreshold
    })

    const itemStyles = computed(() => {
      const styles: StyleObject = {}

      styles.padding = parseGutter(props.itemGutter)
        .reverse()
        .map(v => `${v}px`)
        .join(' ')

      return styles
    })

    const enablePullDirections = computed<PullDirection[]>(() => {
      if (props.enablePullRefresh) {
        return horizontal ? ['right'] : ['down']
      }

      return []
    })

    watch(
      () => props.data,
      val => {
        dataToList(val)
        nextTick(() => updateItems(null, 'dataChange'))
      },
      {
        deep: true
      }
    )

    watch(
      () => props.itemSize,
      () => updateItems(null, 'itemSizeChange')
    )

    dataToList(props.data)

    useResizeDetector(scrollView, onResize)

    onMounted(() => {
      updateSize()
      updateItems(null, 'init')
    })

    return {
      listEl,
      scrollView,
      scrollX,
      scrollY,
      horizontal,
      list,
      cols,
      lowerThreshold,
      enablePullDirections,
      itemStyles,
      onRefreshing,
      recordInteraction,
      onScroll,
      onScrollToLower,
      scrollTo: scrollToOffset,
      scrollToIndex,
      scrollToEnd,
      locale
    }
  }
})
</script>
