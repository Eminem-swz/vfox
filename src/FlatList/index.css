.fx-flat-list {
  position: relative;
  display: block;
  width: 100%;
}

.fx-flat-list_list {
  list-style: none;
  margin: 0;
  padding: 0;
  position: relative;
  -webkit-transform: translateZ(0);
  transform: translateZ(0);
}

.fx-flat-list_item {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-orient: vertical;
  -webkit-box-direction: normal;
  -ms-flex-direction: column;
  flex-direction: column;
  left: 0;
  top: 0;
}

.fx-flat-list_item-inner {
  height: 100%;
  -webkit-box-flex: 1;
  -ms-flex-positive: 1;
  flex-grow: 1;
}

.fx-flat-list.horizontal .fx-flat-list_header, .fx-flat-list.horizontal .fx-flat-list_footer {
  display: inline-block;
  vertical-align: top;
}

.fx-flat-list.horizontal {
  white-space: nowrap;
}

.fx-flat-list.horizontal .fx-flat-list_list {
  display: -webkit-inline-box;
  display: -ms-inline-flexbox;
  display: inline-flex;
  height: 100%;
}

.fx-flat-list.horizontal .fx-flat-list_item {
  display: -webkit-inline-box;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-orient: horizontal;
  -webkit-box-direction: normal;
  -ms-flex-direction: row;
  flex-direction: row;
  height: 100%;
  vertical-align: top;
}

.fx-flat-list_indicator {
  display: -webkit-box;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  justify-content: center;
  width: 100%;
  font-size: 14px;
  color: #595959;
  -webkit-box-sizing: border-box;
  box-sizing: border-box;
  height: 48px;
}

.fx-flat-list_indicator .fx-icon {
  --size: 18px;
  --color: #595959;
  margin-right: 8px;
}

.fx-flat-list_indicator span {
  line-height: 16px;
  text-align: center;
  white-space: normal;
}
