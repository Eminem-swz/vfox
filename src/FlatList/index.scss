@import '../style/var.scss';

.#{$prefix}-flat-list {
  position: relative;
  display: block;
  width: 100%;

  &_list {
    list-style: none;
    margin: 0;
    padding: 0;
  }

  &_item {
    display: flex;
    flex-direction: column;

    &-inner {
      height: 100%;
      flex-grow: 1;
    }
  }

  &_header,
  &_footer {
    .#{$prefix}-flat-list.horizontal & {
      display: inline-block;
      vertical-align: top;
    }
  }

  &.horizontal {
    white-space: nowrap;

    .#{$prefix}-flat-list {
      &_list {
        display: inline-flex;
        height: 100%;
      }

      &_item {
        display: inline-flex;
        flex-direction: row;
        height: 100%;
        vertical-align: top;
      }
    }
  }

  &_indicator {
    display: flex;
    align-items: center;
    justify-content: center;
    width: 100%;
    font-size: 14px;
    color: $font-color;
    box-sizing: border-box;
    height: 48px;

    .#{$prefix}-icon {
      --size: 18px;
      --color: #{$font-color};
      margin-right: 8px;
    }

    span {
      line-height: 16px;
      text-align: center;
      white-space: normal;
    }
  }
}
