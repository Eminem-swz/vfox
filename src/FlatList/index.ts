import { App } from 'vue'
import { SFCWithInstall } from '@/utils/types'
import FlatList from './FlatList.vue'

const _FlatList: SFCWithInstall<typeof FlatList> = Object.assign(FlatList, {
  install: function(app: App) {
    app.component(FlatList.name, FlatList)
  }
})

export default _FlatList
