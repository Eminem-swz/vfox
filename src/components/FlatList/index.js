import FlatList from './FlatList.vue'

FlatList.install = function(Vue) {
  Vue.component(FlatList.name, FlatList)
}

export default FlatList
