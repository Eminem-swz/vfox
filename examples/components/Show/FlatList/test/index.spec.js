/* global describe,test,expect */

import { shallowMount } from '@vue/test-utils'
import FlatList from '@/FlatList'

const data = []

for (let i = 0; i < 10; i++) {
  data.push({
    id: i + 1,
    text: `Item ${i + 1}.`
  })
}

describe('FlatList', () => {
  test('should render default slot correctly', () => {
    const wrapper = shallowMount(FlatList, {
      propsData: {
        data
      },
      scopedSlots: {
        item: `<div>{{ props.item.text }}</div>`
      }
    })

    expect(wrapper.html()).toMatchSnapshot()
  })
})
