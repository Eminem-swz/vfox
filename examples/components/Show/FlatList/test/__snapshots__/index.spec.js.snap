// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlatList should render default slot correctly 1`] = `<fx-scroll-view-stub class="fx-flat-list"><object style="display: block; position: absolute; top: 0px; left: 0px; width: 100%; height: 100%; padding: 0px; margin: 0px; opacity: 0; z-index: -1000; pointer-events: none;" tabindex="-1" type="text/html" aria-hidden="true" data="about:blank"></object></fx-scroll-view-stub>`;
